src/input.cpp

```diff
@@ -1145,38 +1157,205 @@ std::string input_context::get_desc(
         }
     }
 
+    std::string padded_text = text;
+    if( n_text > 0 ) {
+        padded_text = string_format( "\u00A0%s", padded_text );
+    }
     if( na ) {
         //~ keybinding description for unbound or non-applicable keys
-        return string_format( separate_fmt, pgettext( "keybinding", "n/a" ), text );
-    } else {
-        return string_format( separate_fmt, get_desc( action_descriptor, 1, evt_filter ), text );
+        return string_format( separate_fmt, pgettext( "keybinding", "n/a" ), padded_text );
     }
+    return string_format( separate_fmt, get_desc( action_descriptor, 1, evt_filter ), padded_text );
 }
 
 std::string input_context::get_desc(
     const std::string &action_descriptor,
     const std::string &text,
-    const input_event_filter &evt_filter ) const
+    const keybinding_hint_state state,
+    const bool show_separators,
+    const input_event_filter &evt_filter
+) const
+{
+    translation inline_fmt;
+    translation separate_fmt;
+    std::string lb_sep;
+    std::string rb_sep;
+    std::string lp_sep;
+    std::string rp_sep;
+    if( show_separators ) {
+        lb_sep = get_option<std::string>( "KEYBINDING_HINT_SEPARATE_SEPARATOR_LEFT" );
+        rb_sep = get_option<std::string>( "KEYBINDING_HINT_SEPARATE_SEPARATOR_RIGHT" );
+        lp_sep = get_option<std::string>( "KEYBINDING_HINT_INLINE_SEPARATOR_LEFT" );
+        rp_sep = get_option<std::string>( "KEYBINDING_HINT_INLINE_SEPARATOR_RIGHT" );
+    }
+    std::string c_key = get_tag_from_color( get_hint_color_for_key( state ) );
+    if( state != keybinding_hint_state::NONE ) {
+        std::string c_sep = get_tag_from_color( get_hint_color_for_separator( state ) );
+        std::string c = get_tag_from_color( get_hint_color( state ) );
+        inline_fmt = to_translation( "keybinding",
+                                     string_format( "%s%s</color>%s%s</color>%s%s</color>%s%s</color>%s%s</color>", c, "%1$s", c_sep,
+                                             lp_sep, c_key, "%2$s", c_sep, rp_sep, c, "%3$s" ) );
+        separate_fmt = to_translation( "keybinding",
+                                       string_format( "%s%s</color>%s%s</color>%s%s</color>%s%s</color>", c_sep, lb_sep, c_key, "%1$s",
+                                               c_sep, rb_sep, c, "%2$s" ) );
```

Comment:
The arguments to `to_translation` have to be string literals or it won't work (the string extractor won't know what runtime values the format arguments will take).

Also formatting a string with format specifiers as arguments to form a new format string feels too hacky to me, and it will likely break if some of the arguments happen to contain a sequence that can be interpreted as a format specifier.

I suggest passing fixed format strings and the needed variables as arguments to the final `get_desc` call instead and format the text in one go there.

Might want to use `colorize` instead of color tag string too to avoid confusing the translators with unpaired color tags.

Reply:
Weird, I responded to you this morning but I guess I never hit return? AHHHHHHH

This code was absolutely atrocious :) I refactored it in the way you suggested, thanks!